openmls/tests/book_code.rs

```diff
@@ -1391,3 +1391,107 @@ fn test_empty_input_errors(ciphersuite: Ciphersuite, provider: &impl OpenMlsProv
+        .leaf_node_capabilities(capabilities.clone())
+        .build(ciphersuite, provider, &bob_signer, bob_credential_with_key)
+        .unwrap();
+
+    // Create a group that supports the custom proposal type
+    let mut alice_group = MlsGroup::builder()
+        .with_capabilities(capabilities.clone())
+        .ciphersuite(ciphersuite)
+        .build(provider, &alice_signer, alice_credential_with_key)
+        .unwrap();
+    // ANCHOR_END: custom_proposal_type
+
+    // Add Bob
+    let (_mls_message, welcome, _group_info) = alice_group
+        .add_members(provider, &alice_signer, &[bob_key_package])
+        .unwrap();
+
+    alice_group.merge_pending_commit(provider).unwrap();
+
+    let staged_welcome = StagedWelcome::new_from_welcome(
+        provider,
+        &MlsGroupJoinConfig::default(),
+        welcome.into_welcome().unwrap(),
+        Some(alice_group.export_ratchet_tree().into()),
+    )
+    .unwrap();
+
+    let mut bob_group = staged_welcome.into_group(provider).unwrap();
+
+    // ANCHOR: custom_proposal_usage
+    // Create a custom proposal based on an example payload and the custom
+    // proposal type defined above
+    let custom_proposal_payload = vec![0, 1, 2, 3];
+    let custom_proposal = CustomProposal((custom_proposal_type, custom_proposal_payload.clone()));
+
+    let (custom_proposal_message, _proposal_ref) = alice_group
+        .propose_custom_proposal_by_reference(provider, &alice_signer, custom_proposal.clone())
+        .unwrap();
+
+    // Have bob process the custom proposal.
+    let processed_message = bob_group
+        .process_message(
+            provider,
+            custom_proposal_message.into_protocol_message().unwrap(),
+        )
+        .unwrap();
+
+    let ProcessedMessageContent::ProposalMessage(proposal) = processed_message.into_content()
+    else {
+        panic!("Unexpected message type");
+    };
+
+    bob_group.store_pending_proposal(*proposal);
+
+    // Commit to the proposal
+    let (commit, _, _) = alice_group
+        .commit_to_pending_proposals(provider, &alice_signer)
+        .expect("Error creating commit");
+
+    let processed_message = bob_group
+        .process_message(provider, commit.into_protocol_message().unwrap())
+        .expect("Error processing commit");
+
+    let staged_commit = match processed_message.into_content() {
+        ProcessedMessageContent::StagedCommitMessage(staged_commit) => staged_commit,
+        _ => panic!("Unexpected message type"),
+    };
+
+    // Check that the proposal is present in the staged commit
+    let expected_proposal = OtherProposal(custom_proposal_payload.into());
+    assert!(staged_commit.queued_proposals().any(|qp| {
+        let Proposal::Other(custom_proposal) = qp.proposal() else {
+            return false;
+        };
+        custom_proposal.0 == custom_proposal_type && custom_proposal.1 == expected_proposal
+    }));
+
+    // ANCHOR_END: custom_proposal_usage
+}
```

Comment:
Just a nit, but the code uses both `expect()` and `unwrap()` arbitrarily. I think `unwrap()` is fine when it is self-explanatory, we have used it in the past.